Fix for #4136 connection to linux refused by using wrong Xauthority file

awalon opened the pull request:
Solution:
1. Get file from sddm-greeter and startplasma-x11 (aka KDE) processes as well
2. Get users home dir from system instead of hard coded pattern /home/<user>
3. Prefer XAUTHORITY environment variable over Xorg -auth if available

awalon committed:
Fix for #4136 connection to linux refused by using wrong Xauthority file
Solution:
1. Get file from sddm-greeter and startplasma-x11 (aka KDE) processes as well
2. Get users home dir from system instead of hard coded pattern /home/<user>
3. Prefer XAUTHORITY environment variable over Xorg -auth if available

rustdesk commented on libs/hbb_common/src/config.rs:
```diff
@@ -322,7 +322,19 @@ fn patch(path: PathBuf) -> PathBuf {
                     if user != "root" {
-                        return format!("/home/{user}").into();
+                        let cmd = format!("getent passwd '{}' | awk -F':' '{{print $6}}'", user);
+                        if let Ok(output) = std::process::Command::new(cmd).output() {
```
How about if here is not OK?

awalon replied:
Good point missed that case, will be fixed by followup PR

rustdesk commented on src/platform/linux.rs:
```diff
@@ -449,12 +451,37 @@ pub fn get_active_username() -> String {
+    return match get_user_by_name(username) {
+        None => {
+            log::debug!("no uid found for user: {}", username);
+            None }
```
format

awalon replied:
Missing line breake (after 2nd None) by updated PR as well

rustdesk commented on src/platform/linux.rs:
```diff
@@ -824,56 +862,107 @@ mod desktop {
-                                    if home_dir.is_empty() {
-                                        self.xauth = format!("/home/{}/{}", self.username, v);
+
+                    if let Some(pid) = line.split_whitespace().nth(1) {
```
Why add one more `if` here?

awalon replied:
Thanks for review, that's a left over for Xauth file fallback. Cann t test this here so this step will be  removed again with next PR.

rustdesk commented on src/platform/linux.rs:
```diff
@@ -824,56 +862,107 @@ mod desktop {
+                                        };
                                     } else {
-                                        self.xauth = format!("{}/{}", home_dir, v);
+                                        // unreachable!
```
Can you explain what changes here line870 -> line 895?

awalon replied:
Highlevel: 
As mentioned by newly added debug messages auth- file will be extraxcted from Xorg process and for a relative path home path will be added.

If no HOME var was found the home folder for given user will be determined from system config (new function: get_user_home_by_name)  old logic just used hard coded  default pattern "/home/<username>" (now only if no system config was found)  which wasn't working for domain users (AD, samba, freeipa, ...) where an additional domain subfolder (/home/<domain>/<username>) needs to be added.

In addition detected file needs to exist, if not this loop won't detect any auth- file. So next step can detect a valid file.

Next step (Todo): Detect auth file by checking process tree (processes below Xorg) without need of predefined checks for Gname, XFCE, Plasma, ...). But as this was my first modifiaction using rust language, there was too less time for better solution.

awalon committed:
Merge branch 'rustdesk:master' into master

awalon committed:
Fix for #4136 connection to linux refused by using wrong Xauthority file.
This PR handles comments of review for PR #4165.

awalon committed:
Merge remote-tracking branch 'origin/master'

rustdesk commented on src/platform/linux.rs:
```diff
@@ -824,22 +863,36 @@ mod desktop {
                 "ps -u {} -f | grep 'Xorg' | grep -v 'grep'",
                 &self.uid
             )) {
+                log::debug!("extracting xauth file for Xorg running with uid: {}", self.uid);
```
Too many log::debug! which it will be print to file, We'd better remove these log::debug before code merge.

awalon replied:
Debug messages was removed now

awalon committed:
Fix for #4136 connection to linux refused by using wrong Xauthority file.
This PR handles comments of 2nd review for initial PR #4165.
Reduced number of info/log messages

awalon commented on libs/hbb_common/src/config.rs:
```diff
@@ -322,6 +322,17 @@ fn patch(path: PathBuf) -> PathBuf {
+                                .trim()
+                                .to_owned();
+                            if !home_dir.is_empty() {
+                                log::info!("config::patch: got home dir from: {}", home_dir);
```
log message removed

rustdesk committed:
Update linux.rs

rustdesk commented:
#4136

rustdesk commented:
I guess https://github.com/rustdesk/rustdesk/issues/440 has the same, no idea what process it is.

rustdesk commented:
#440
